File: cses/06_tree_algorithms/1687_company-queries-i/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần để tăng tốc
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    LOG = 18  # 2^18 = 262144 > 2*10^5 >= n

    # parent[v] = cha trực tiếp của v (1-indexed); parent[1] = 0 vì gốc không có cha
    parent = [0] * (n + 1)
    for v in range(2, n + 1):
        parent[v] = int(data[idx]); idx += 1

    # Binary lifting lưu dạng mảng phẳng: up[j*size + v] = tổ tiên cách v đúng 2^j bậc
    # (node ảo 0 hấp thụ mọi bước nhảy vượt gốc)
    size = n + 1
    up = [0] * (LOG * size)

    # Tầng 0 chính là cha trực tiếp
    for v in range(1, n + 1):
        up[v] = parent[v]

    # Tầng cao dựng theo công thức 2^j = 2^(j-1) + 2^(j-1)
    for j in range(1, LOG):
        base = j * size
        prev = (j - 1) * size
        for v in range(1, n + 1):
            up[base + v] = up[prev + up[prev + v]]

    out = []
    for _ in range(q):
        x = int(data[idx]); k = int(data[idx + 1]); idx += 2
        # Nhảy theo từng bit được bật của k; dừng sớm nếu x đã rơi về node ảo 0
        j = 0
        kk = k
        while kk and x:
            if kk & 1:
                x = up[j * size + x]
            kk >>= 1
            j += 1
        # x == 0 nghĩa là tổ tiên không tồn tại -> in -1
        out.append(str(x) if x else "-1")

    sys.stdout.buffer.write("\n".join(out).encode())
    sys.stdout.buffer.write(b"\n")


if __name__ == "__main__":
    main()

File: cses/06_tree_algorithms/1687_company-queries-i/solution.cpp
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::size_t n, query_count;
    std::cin >> n >> query_count;

    // Chọn số tầng LOG sao cho 2^levels > n, đủ biểu diễn mọi k <= n
    std::size_t levels = 1;
    while ((std::size_t{1} << levels) <= n) {
        ++levels;
    }

    // Bảng binary lifting: ancestor[j][v] = tổ tiên cách v đúng 2^j bậc
    // (node ảo 0 dùng cho trường hợp "rơi" ra ngoài cây)
    std::vector<std::vector<std::size_t>> ancestor(
        levels, std::vector<std::size_t>(n + 1, 0));

    // Tầng 0 chính là cha trực tiếp; ancestor[0][1] = 0 vì gốc không có cha
    for (std::size_t employee = 2; employee <= n; ++employee) {
        std::cin >> ancestor[0][employee];
    }

    // Xây các tầng cao theo công thức 2^j = 2^(j-1) + 2^(j-1)
    for (std::size_t level = 1; level < levels; ++level) {
        for (std::size_t employee = 1; employee <= n; ++employee) {
            ancestor[level][employee] =
                ancestor[level - 1][ancestor[level - 1][employee]];
        }
    }

    while (query_count-- > 0) {
        std::size_t employee, steps;
        std::cin >> employee >> steps;

        // Nhảy theo từng bit được bật của steps; dừng sớm nếu đã rơi về node ảo 0
        for (std::size_t level = 0; level < levels && employee != 0; ++level) {
            if ((steps & (std::size_t{1} << level)) != 0) {
                employee = ancestor[level][employee];
            }
        }

        // employee == 0 nghĩa là tổ tiên không tồn tại
        if (employee == 0) {
            std::cout << -1 << '\n';
        } else {
            std::cout << employee << '\n';
        }
    }
}
